# 07_rag_evaluation: MRR and NDCG Metrics

This notebook implements mathematical search evaluations: Hit Rate, Mean Reciprocal Rank (MRR), and Normalized Discounted Cumulative Gain (NDCG).


In [1]:
import numpy as np

def compute_mrr(retrieved_ranks):
    reciprocal_ranks = [1.0 / r if r > 0 else 0.0 for r in retrieved_ranks]
    return np.mean(reciprocal_ranks)

def compute_dcg(relevance_scores):
    scores = np.array(relevance_scores)
    discounts = np.log2(np.arange(len(scores)) + 2)
    return np.sum(scores / discounts)

def compute_ndcg(relevance_scores):
    actual_dcg = compute_dcg(relevance_scores)
    ideal_scores = sorted(relevance_scores, reverse=True)
    ideal_dcg = compute_dcg(ideal_scores)
    return actual_dcg / ideal_dcg if ideal_dcg > 0 else 0.0

query_ranks = [2, 1, 4]
mrr = compute_mrr(query_ranks)

actual_relevance = [2, 3, 0]
ndcg = compute_ndcg(actual_relevance)

print(f"Mean Reciprocal Rank (MRR): {mrr:.4f}")
print(f"NDCG@3 Score: {ndcg:.4f}")


Mean Reciprocal Rank (MRR): 0.5833
NDCG@3 Score: 0.9134


### Output Explanation
- MRR prints as $0.5833$, and NDCG@3 evaluates to $0.9134$, confirming our theoretical formulas.
